src/backend/distributed/metadata/node_metadata.c

```diff
@@ -874,14 +1524,25 @@ SyncDistributedObjectsToNodeList(List *workerNodeList)
 
 	Assert(ShouldPropagate());
 
-	List *commandList = SyncDistributedObjectsCommandList(workerNode);
+	/* send systemwide objects e.g. roles */
+	SendNodeWideObjectsSyncCommands(workerNodeList);
```

Comment:
hmm, the test UDF `activate_node_snapshot` still uses `SyncDistributedObjectsCommandList`. I wish we had a way for that to also use the new functions. Because, now it doesn't make much sense to have the test UDF as it doesn't rely on the actual functions.

besides, we probably wouldn't want to keep such large functions (i.e., SyncDistributedObjectsCommandList/PgDistTableMetadataSyncCommandList etc) in the Citus source code.

So, I wonder if we could change the design a little bit. We introduce `MetadataSyncContext` in this commit, and pass that to functions like `SyncDistributedObjectsCommandList` etc. 

In the functions, either collect the commands (for the test UDF), or immediately send to the workers. That'd avoid us to write new functions like `SendIterableDetachPartitionCommands` etc.

Do you think is that possible to re-work the patch like that?

Reply:
I need to have a quick look at for refactor. For draft purpose, I created separate functions for the new nontransactional metadata sync but I agree it would be nicer to use the same methods by making use of metadatacontext as much as possible.  Let me check and update.

Reply:
Refactored so that both coordinated and nontransactional modes uses the same API now. We have a problem revealed by the failure test of nontransactional mode: distributed object metadata depends on node metadata. But when I move node metadata sync before distributed object sync, coordinated transaction just hangs. I need to figure it out. fyi

Reply:
We have 1 test failure at `upgrade_post_11_after.sql`. It compares metadata commands between coordinator and workers and expects 0 difference. But now we send `SELECT worker_drop_shell_table($$shell_table_name$$);` by looking at pg_dist_partition at coordinator node. There is a difference like below

```sql
- (0 rows)
+ SELECT worker_drop_shell_table($$upgrade_basic.t_range$$);
+(1 row)
```

which indicates coordinator has some entry in pg_dist_partition which is missing at workers. (I donot know how that is missing, I think it is not supposed to be missing)

Reply:
refactored by 3. commit. 
We added a procedure which accepts a flag and drops all shell tables in single transaction if the flag is true. Otherwise, it creates a new transaction per shell table during shell table drops.